File: problems-extend-adapt/0701-0800/0711_unique-island-shapes-ii/solution.py
from typing import List, Set, Tuple


class Solution:
    def countUniqueShapes(self, grid: List[List[int]]) -> int:
        # Flood-fill each island with an explicit stack, then name the shape by
        # the lexicographically smallest normalized cell list among its eight
        # rotations and reflections: two islands share a name exactly when one
        # maps onto the other under the statement's rule.
        m, n = len(grid), len(grid[0])
        seen = [[False] * n for _ in range(m)]
        shapes: Set[Tuple[Tuple[int, int], ...]] = set()
        for i in range(m):
            for j in range(n):
                if grid[i][j] != 1 or seen[i][j]:
                    continue
                seen[i][j] = True
                stack: List[Tuple[int, int]] = [(i, j)]
                cells: List[Tuple[int, int]] = []
                while stack:
                    r, c = stack.pop()
                    cells.append((r, c))
                    for nr, nc in ((r - 1, c), (r + 1, c), (r, c - 1), (r, c + 1)):
                        if 0 <= nr < m and 0 <= nc < n and grid[nr][nc] == 1 and not seen[nr][nc]:
                            seen[nr][nc] = True
                            stack.append((nr, nc))
                signs = (1, -1)
                best: Tuple[Tuple[int, int], ...] = ()
                for t in range(8):
                    a, b, swap = signs[t & 1], signs[(t >> 1) & 1], t & 4 != 0
                    moved = [(a * (c if swap else r), b * (r if swap else c)) for r, c in cells]
                    r0 = min(row for row, _ in moved)
                    c0 = min(col for _, col in moved)
                    shape = tuple(sorted((row - r0, col - c0) for row, col in moved))
                    if not best or shape < best:
                        best = shape
                shapes.add(best)
        return len(shapes)

File: problems-extend-adapt/0701-0800/0711_unique-island-shapes-ii/solution.rs
use std::collections::HashSet;

impl Solution {
    // Flood-fill each island with an explicit stack, then name the shape by
    // the lexicographically smallest normalized cell list among its eight
    // rotations and reflections: two islands share a name exactly when one
    // maps onto the other under the statement's rule.
    pub fn count_unique_shapes(grid: Vec<Vec<i32>>) -> i32 {
        let m = grid.len();
        let n = grid[0].len();
        let mut seen = vec![vec![false; n]; m];
        let mut shapes: HashSet<Vec<(i32, i32)>> = HashSet::new();
        let sign = [1i32, -1i32];
        for i in 0..m {
            for j in 0..n {
                if grid[i][j] != 1 || seen[i][j] {
                    continue;
                }
                seen[i][j] = true;
                let mut stack = vec![(i, j)];
                let mut cells: Vec<(usize, usize)> = Vec::new();
                while let Some((r, c)) = stack.pop() {
                    cells.push((r, c));
                    for (dr, dc) in [(-1isize, 0isize), (1, 0), (0, -1), (0, 1)] {
                        let nr = r as isize + dr;
                        let nc = c as isize + dc;
                        if nr < 0 || nr >= m as isize || nc < 0 || nc >= n as isize {
                            continue;
                        }
                        let (nr, nc) = (nr as usize, nc as usize);
                        if grid[nr][nc] == 1 && !seen[nr][nc] {
                            seen[nr][nc] = true;
                            stack.push((nr, nc));
                        }
                    }
                }
                let mut best: Vec<(i32, i32)> = Vec::new();
                for t in 0..8 {
                    let (a, b, swap) = (sign[t & 1], sign[(t >> 1) & 1], t & 4 != 0);
                    let mut moved: Vec<(i32, i32)> = cells
                        .iter()
                        .map(|&(r, c)| {
                            let nr = a * if swap { c as i32 } else { r as i32 };
                            let nc = b * if swap { r as i32 } else { c as i32 };
                            (nr, nc)
                        })
                        .collect();
                    let r0 = moved.iter().map(|&(r, _)| r).min().unwrap();
                    let c0 = moved.iter().map(|&(_, c)| c).min().unwrap();
                    for cell in &mut moved {
                        cell.0 -= r0;
                        cell.1 -= c0;
                    }
                    moved.sort();
                    if best.is_empty() || moved < best {
                        best = moved;
                    }
                }
                shapes.insert(best);
            }
        }
        shapes.len() as i32
    }
}
